gui: Add close window shortcut

IPGlider opened the pull request:
CMD+W is the standard shortcut in macOS to close a window without
exiting the program.

This adds support to use the shortcut in both main and debug windows.

IPGlider committed:
gui: Add close window shortcut

CMD+W/CTRL+W is the standard shortcut to close a window without
exiting the program.

jonasschnelli commented:
Thanks!
Tested ACK 20859aef8b571542772a42b4c413c4a71a6576c5

The macOS HIG about Command+W:
https://developer.apple.com/design/human-interface-guidelines/macos/user-interaction/keyboard/

promag commented on src/qt/bitcoingui.cpp:
```diff
@@ -409,6 +409,10 @@ void BitcoinGUI::createActions()
     connect(new QShortcut(QKeySequence(Qt::CTRL + Qt::SHIFT + Qt::Key_D), this), &QShortcut::activated, this, &BitcoinGUI::showDebugWindow);
+
+#ifdef Q_OS_MAC
+    connect(new QShortcut(QKeySequence(Qt::CTRL + Qt::Key_W), this), &QShortcut::activated, this, [this]() { this->hide(); });
```
No need to lambda, could just be `... , &QShortcut::activated, this, &QWidget::hide)`?

promag reviewed (commented):
Pressing ESC also closes the dialogs (except the main window). Is this something to keep in macos? (not suggesting it though, just want to mention the redundancy)

laanwj commented:
utACK, ideally this kind of stuff would be handled by the cross-platform windowing system, so it's a bit disappointing that it's not, but also, this is only a few lines and increases user friendliness.

BTW: CTRL-W is also quite a common shortcut for closing windows in applications outside of OS-X, for example in browsers it closes the tab.

IPGlider commented:
I will do the suggested changes and add the functionality to all the windows.

Thanks for the review.

laanwj commented:
`QShortcut` has a `Qt::ApplicationShortcut` for application-global shortcuts (see https://doc.qt.io/Qt-5/qt.html#ShortcutContext-enum)—would this be useful here?

luke-jr commented:
Any reason not to enable this on all platforms?

IPGlider commented:
I have updated the PR with the given suggestions and added the shortcut to most windows.

The only one without this shortcut is the 'About Qt' as it is part of Qt and I have not been able to add the shortcut to it.

If I have missed any other window please let me know.

@laanwj `Qt::ApplicationShortcut` does not work for the intended purpose. It makes the shortcut works even if it is not focused.


@luke-jr CMD+W is usually only used on macOS, more information can be found in the [macOS HID](https://developer.apple.com/design/human-interface-guidelines/macos/user-interaction/keyboard/) posted by @jonasschnelli and in [Wikepedia's Table of keyboard shortcuts](https://en.wikipedia.org/wiki/Table_of_keyboard_shortcuts).

luke-jr commented:
I use Ctrl+W on Linux all the time to close browser tabs. My IRC client seems to use it for closing tabs too.

Furthermore, if macOS is using it, we *can't* use it for anything else (eg, closing wallets?).

So overall, I'd say it makes sense to just support it on all platforms.

luke-jr commented:
Note: When extending it to support other platforms, care needs to be taken that the main window is not actually hidden if there's no system tray icon. Probably should minimise in that case? Not sure...

luke-jr commented on src/qt/utilitydialog.cpp:
```diff
@@ -158,6 +160,8 @@ ShutdownWindow::ShutdownWindow(QWidget *parent, Qt::WindowFlags f):
         tr("Do not shut down the computer until this window disappears.")));
     setLayout(layout);
+
+    GUIUtil::handleHideShortcut(this);
```
Do we want this to be hide-able??

IPGlider replied:
The criteria I followed to make windows hide-able has been what I understand is the common usage in macOS, meaning that every window that has a close button can also be closed using the CMD+W shortcut.

luke-jr commented:
Probably a number of windows ought to be closed normally, not merely hidden?

jonasschnelli commented on src/qt/guiutil.cpp:
```diff
@@ -370,6 +371,13 @@ void bringToFront(QWidget* w)
+void handleHideShortcut(QWidget* w)
+{
+#ifdef Q_OS_MAC
+    w->connect(new QShortcut(QKeySequence(Qt::CTRL + Qt::Key_W), w), &QShortcut::activated, w, &QWidget::hide);
```
I haven't looked closely at this,... but is there a proper deallocation of that `QShortcut`?

IPGlider replied:
I have followed the official documentation in this regard[0]. If you consider that this is not a correct application please let me know a more appropriate way.

[0] https://doc.qt.io/qt-5/qobject.html

hebasto commented:
@IPGlider Could you address reviewers' comments?

IPGlider commented:
Is working in this PR still worth it? I would like to finish it.

@hebasto I will address the comments in a few days.

hebasto commented:
@IPGlider 
> Is working in this PR still worth it?

Yes, [IMO](https://github.com/bitcoin/bitcoin/pull/16515#issuecomment-517917203).

IPGlider commented:
@luke-jr 

> I use Ctrl+W on Linux all the time to close browser tabs. My IRC client seems to use it for closing tabs too.
>
> Furthermore, if macOS is using it, we can't use it for anything else (eg, closing wallets?).
>
> So overall, I'd say it makes sense to just support it on all platforms.

I wasn't aware that this shortcut was also commonly used in other OSs other than macOS. Considering this I will remove the ifdef to support all OSs.

> Probably a number of windows ought to be closed normally, not merely hidden?

I agree and after reading the QT documentation I will replace the `hide` call with a `close` call.


Should I continue working on this PR considering that the name of the branch is `add_cmd_w_support_in_macos` or would it be preferable to create another PR?

fanquake commented:
@IPGlider No need for a new PR. Just update the PR title, body etc if required.

IPGlider commented:
Two jobs failed in Travis CI during the set up:

- https://travis-ci.org/bitcoin/bitcoin/jobs/568415330
- https://travis-ci.org/bitcoin/bitcoin/jobs/568415332

I have not found a way to force a re-run. Nevertheless the other jobs completed successfully.

hebasto commented:
@IPGlider 
> Two jobs failed in Travis CI...

Restarted...

hebasto commented on src/qt/guiutil.cpp:
```diff
@@ -378,6 +379,11 @@ void bringToFront(QWidget* w)
 
+void handleCloseWindowShortcut(QWidget* w)
+{
+    w->connect(new QShortcut(QKeySequence(Qt::CTRL + Qt::Key_W), w), &QShortcut::activated, w, &QWidget::close);
```
`QObject::connect()` is a static function of `QObject`:
```suggestion
    QObject::connect(new QShortcut(QKeySequence(Qt::CTRL + Qt::Key_W), w), &QShortcut::activated, w, &QWidget::close);
```

GChuf commented:
@IPGlider are you still working on this?

jonasschnelli commented on src/qt/bitcoingui.cpp:
```diff
@@ -210,6 +210,8 @@ BitcoinGUI::BitcoinGUI(interfaces::Node& node, const PlatformStyle *_platformSty
     m_app_nap_inhibitor = new CAppNapInhibitor;
 #endif
+
+    GUIUtil::handleCloseWindowShortcut(this);
```
I think it should also close the main window (at least on macOS it is the case for almost all apps).

IPGlider replied:
The CTRL+W/CMD+W behaviour in macOS is usually to close the focused window.

jonasschnelli commented:
Needs rebase. I think this is almost done. Let's finish it @IPGlider.

GChuf commented:
IPglider hasn't been active since April I think - (when) is it okay if I try to pick it up and include hebasto's suggestions?

IPGlider commented:
Sorry for the delay, I will rebase and continue working on this PR this weekend.

GChuf commented:
Excellent, good luck!

IPGlider commented on src/qt/splashscreen.cpp:
```diff
@@ -8,6 +8,7 @@
 
 #include <qt/splashscreen.h>
 
+#include <qt/guiutil.h>
```
The rationale for adding the shortcut here is that on every window with an enabled close button the shortcut should be applicable.

It could be worth it taking into consideration disabling the close button or removing the bar all together.

IPGlider commented on src/qt/utilitydialog.cpp:
```diff
@@ -114,6 +115,8 @@ HelpMessageDialog::HelpMessageDialog(interfaces::Node& node, QWidget *parent, bo
         ui->aboutLogo->setVisible(false);
     }
+
+    GUIUtil::handleCloseWindowShortcut(this);
```
Which of the following cases is preferable?

Two windows with their close behaviour consistent among them but inconsistent with the behaviour of the rest of them or keeping the inconsistency only on the QT window?

IPGlider commented on src/qt/utilitydialog.cpp:
```diff
@@ -153,6 +156,8 @@ ShutdownWindow::ShutdownWindow(QWidget *parent, Qt::WindowFlags f):
         tr("Do not shut down the computer until this window disappears.")));
     setLayout(layout);
+
+    GUIUtil::handleCloseWindowShortcut(this);
```
Same comment as for the splash screen.

hebasto reviewed (approved):
ACK f5a3a5b9ab362c58fa424261f313aa9cf46d2a98, tested on Linux Mint 19.3 by manually opening available dialogs and sub-windows, and applying the `Ctrl+W` shortcut. Also tested with "Minimize on close" option enabled / disabled.

jonasschnelli commented:
Tested ACK f5a3a5b9ab362c58fa424261f313aa9cf46d2a98
